bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/TypeChecker.java

```diff
@@ -1279,32 +1297,118 @@ private static boolean checkIsArrayType(BType sourceType, BArrayType targetType,
     }
 
     private static boolean checkIsTupleType(BType sourceType, BTupleType targetType, List<TypePair> unresolvedTypes) {
-        if (sourceType.getTag() != TypeTags.TUPLE_TAG) {
-            return false;
+        int sourceTypeTag = sourceType.getTag();
+
+        if (sourceTypeTag == TypeTags.UNION_TAG) {
+            for (BType memberType : ((BUnionType) sourceType).getMemberTypes()) {
+                if (!checkIsTupleType(memberType, targetType, unresolvedTypes)) {
+                    return false;
+                }
+            }
+            return true;
         }
 
-        List<BType> sourceTypes = new ArrayList<>(((BTupleType) sourceType).getTupleTypes());
-        BType sourceRestType = ((BTupleType) sourceType).getRestType();
-        if (sourceRestType != null) {
-            sourceTypes.add(sourceRestType);
+        if (sourceTypeTag != TypeTags.ARRAY_TAG && sourceTypeTag != TypeTags.TUPLE_TAG) {
+            return false;
         }
 
         List<BType> targetTypes = new ArrayList<>(targetType.getTupleTypes());
         BType targetRestType = targetType.getRestType();
-        if (targetRestType != null) {
-            targetTypes.add(targetRestType);
+
+        BTupleType sourceTupleType = null;
+        if (sourceTypeTag == TypeTags.TUPLE_TAG) {
+            sourceTupleType = (BTupleType) sourceType;
+        } else {
+            BArrayType sourceArrayType = (BArrayType) sourceType;
+            BType sourceElementType = sourceArrayType.getElementType();
+
+            switch (sourceArrayType.getState()) {
+                case OPEN:
+                    if (targetRestType == null) {
+                        return false;
+                    }
+                    if (targetTypes.isEmpty()) {
+                        return checkIsType(sourceElementType, targetRestType, unresolvedTypes);
+                    }
+                    return false;
+                case CLOSED:
+                    if (sourceArrayType.getSize() >= targetTypes.size()) {
```

Comment:
We can remove this if-condition, if you bring the condition in the else-if, before this one
i.e:
```java
if (sourceArrayType.getSize() < targetTypes.size()) {
    return false;
}

// rest of the code goes here...
```